src/librustc_mir/transform/generator.rs

```diff
@@ -0,0 +1,777 @@
+                kind: term,
+            }),
+            is_cleanup: true,
+        });
+
+        mir.basic_blocks_mut().push(BasicBlockData {
+            statements: Vec::new(),
+            terminator: Some(Terminator {
+                source_info,
+                kind: TerminatorKind::Resume,
+            }),
+            is_cleanup: true,
+        });
+    }
+
+    cleanup
+}
+
+fn generate_resume<'a, 'tcx>(
+                tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                mut transform: TransformVisitor<'a, 'tcx>,
+                node_id: NodeId,
+                def_id: DefId,
+                source: MirSource,
+                cleanup: Option<BasicBlock>,
+                mir: &mut Mir<'tcx>) {
+    // Poison the generator when it unwinds
+    for block in mir.basic_blocks_mut() {
+        let source_info = block.terminator().source_info;
+        if let &TerminatorKind::Resume = &block.terminator().kind {
+            block.statements.push(transform.set_state(1, source_info));
+        }
+    }
+
+    let source_info = SourceInfo {
+        span: mir.span,
+        scope: ARGUMENT_VISIBILITY_SCOPE,
+    };
+
+    let poisoned_block = BasicBlock::new(mir.basic_blocks().len());
+
+    let term = TerminatorKind::Assert {
+        cond: Operand::Constant(box Constant {
+            span: mir.span,
+            ty: tcx.types.bool,
+            literal: Literal::Value {
+                value: ConstVal::Bool(false),
+            },
+        }),
+        expected: true,
+        msg: AssertMessage::GeneratorResumedAfterPanic,
+        target: transform.return_block,
+        cleanup: cleanup,
+    };
+
+    mir.basic_blocks_mut().push(BasicBlockData {
+        statements: Vec::new(),
+        terminator: Some(Terminator {
+            source_info,
+            kind: term,
+        }),
+        is_cleanup: false,
+    });
+
+    transform.bb_targets.insert((poisoned_block, None), 1);
+
+    let switch = TerminatorKind::SwitchInt {
+        discr: Operand::Consume(transform.make_field(transform.state_field, tcx.types.u32)),
+        switch_ty: tcx.types.u32,
+        values: Cow::from(transform.bb_targets.values().map(|&i| {
+                ConstInt::U32(i)
+            }).collect::<Vec<_>>()),
+        targets: transform.bb_targets.keys()
+            .map(|&(k, _)| k)
+            .chain(once(transform.return_block))
+            .collect(),
+    };
+
+    insert_entry_point(mir, BasicBlockData {
+        statements: Vec::new(),
+        terminator: Some(Terminator {
+            source_info,
+            kind: switch,
+        }),
+        is_cleanup: false,
+    });
+
+    make_generator_state_argument_indirect(tcx, def_id, mir);
+
+    // Make sure we remove dead blocks to remove
+    // unrelated code from the drop part of the function
+    simplify::remove_dead_blocks(mir);
+
+    dump_mir(tcx, None, "generator_resume", &0, source, mir);
+}
+
+fn insert_clean_drop<'a, 'tcx>(mir: &mut Mir<'tcx>) -> (BasicBlock, BasicBlock) {
```

Comment:
This is only for the drop glue, but it has to be done before `elaborate_generator_drops` is called (otherwise we get infinite recursion). `remove_dead_blocks` gets rid of this code in the resume part, so that's not an issue.